File: bin/command/impl/prepare_docker_environment.py
from command.command import Command
from command.command_result import CommandResult
from context.impl.run_chain_context import RunChainContext
from util import docker_utils


class PrepareDockerEnvironment(Command):
    """Prepares the Docker environment; creates volumes, networks."""

    context: RunChainContext

    def __init__(self, context: RunChainContext) -> None:
        self.context = context

    def execute(self) -> CommandResult:
        created_volume = docker_utils.create_volume(self.context.config['DOCKER']['cache_volume'])
        created_network = docker_utils.create_network(self.context.config['DOCKER']['network'])
        return CommandResult.OK if created_volume and created_network else CommandResult.FAILED

File: bin/src/main/java/dev/amrw/bin/chain/command/PrepareDockerEnvironment.java
package dev.amrw.bin.chain.command;

import lombok.extern.log4j.Log4j2;
import org.apache.commons.chain.Command;
import org.apache.commons.chain.Context;

import java.util.List;

/**
 * Prepares the Docker environment; creates volumes, networks.
 */
@Log4j2
public class PrepareDockerEnvironment extends RunChainCommand {

    @Override
    public boolean execute(final Context context) {
        super.prepareContext(context);
        createNetwork();
        createVolume();
        return Command.CONTINUE_PROCESSING;
    }

    private void createNetwork() {
        final var networkName = dockerConfig.getNetwork();
        final var networks = dockerClient.listNetworksCmd().withNameFilter(networkName).exec();
        if (networks.isEmpty()) {
            dockerClient.createNetworkCmd().withName(networkName).exec();
        } else {
            log.info("Network already exists, not creating (name={})", networkName);
        }
    }

    private void createVolume() {
        final var volumeName = dockerConfig.getBuildImageConfig().getVolume();
        final var volumes = dockerClient
                .listVolumesCmd()
                .withFilter("name", List.of(volumeName))
                .exec()
                .getVolumes();
        if (volumes.isEmpty()) {
            dockerClient.createVolumeCmd().withName(volumeName).exec();
        } else {
            log.info("Volume already exists, not creating (name={})", volumeName);
        }
    }
}
